vscode/src/format.ts

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { ILanguageService } from "qsharp-lang";
+import * as vscode from "vscode";
+import { toVscodeRange } from "./common";
+import { EventType, sendTelemetryEvent } from "./telemetry";
+import { getRandomGuid } from "./utils";
+
+export function createFormatProvider(languageService: ILanguageService) {
+  return new QSharpFormatProvider(languageService);
+}
+
+class QSharpFormatProvider implements vscode.DocumentFormattingEditProvider {
+  constructor(public languageService: ILanguageService) {}
+
+  async provideDocumentFormattingEdits(document: vscode.TextDocument) {
+    // telemetry start format
+    const associationId = getRandomGuid();
+    sendTelemetryEvent(EventType.FormatStart, { associationId }, {});
+    const start = performance.now();
+
+    const lsEdits = await this.languageService.getFormatChanges(
+      document.uri.toString(),
+    );
+
+    if (!lsEdits) return [];
```

Comment:
Shouldn't you still send a FormatEnd event if it returns early? Else you have the FormatStart with no FormatEnd whenever this code path is taken.

Reply:
Sending FormatEnd in no lsEdit case.